feat: add support for defining/importing api gateway models

john-shaskin opened the pull request:
Adds support to @aws-cdk/api-gateway module to define one or more models, attached to a given REST API.

**Note:** Did not yet add JSON schema validation for the `schema` property. Toyed with the `jsonschema` Node module, but that wouldn't work, due to the lack of JSII support. Open to some suggestions on this, if this problem has been tackled elsewhere in CDK.

Addresses issue: https://github.com/awslabs/aws-cdk/issues/1695

----

### Pull Request Checklist

* [x] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

RomainMuller commented:
In order to get the `jsonschema` package to work, you'll have to add it to the package's `bundledDependencies`. This way, JSII will not need to have a `.jsii` assembly for it.

RomainMuller commented on packages/@aws-cdk/aws-apigateway/lib/model.ts:
```diff
@@ -40,7 +89,107 @@ export class EmptyModel implements IModel {
+/**
+ * Interface share for creating or importing a model.
+ */
+export interface IModelConstruct extends cdk.IConstruct, IModel {
```
I think `IModel` should be the `IConstruct` sub interface. I'm not sure what you mean when you talk about *simple text reference* on the comment of `IModel` itself...

john-shaskin replied:
When I initially added the `IModel` interface, it was providing support just for the default Empty and Error models that are generated by default for a REST API. In that pull request, the `EmptyModel` and `ErrorModel` types were not set up to implement the `Construct` type, in the interest of keeping things simple. I can try to look into a low-impact way to converge on one IModel interface.

john-shaskin replied:
I've been looking at this, and it seems weird to require the `EmptyModel` and `ErrorModel` classes to extend the full Construct class, when they are just there to be able to return "Empty" and "Error", for method responses using these built-in models.

I can, but it may just mean that I would stop exporting those types, and only expose them as static properties on the `Model` class. If that seems best, I can do it, and then converge things down to one interface.

At the moment, I've changed the naming around:
- `IModelRef` refers to the base interface that just exposes a model ID. Used for (`EmptyModel`, `ErrorModel`)
- `IModel` now refers to the full interface that inherits from `IConstruct`. Used for (`Model` and `ImportedModel`)

RomainMuller commented on packages/@aws-cdk/aws-apigateway/lib/model.ts:
```diff
@@ -40,7 +89,107 @@ export class EmptyModel implements IModel {
-// TODO: Implement Model, enabling management of custom models.
\ No newline at end of file
+/**
+ * Model CFN resource
```
That comment shouldn't be necessary, as the `CfnModel` class would already encapsulate this information.

john-shaskin replied:
Removed the comment.

RomainMuller commented on packages/@aws-cdk/aws-apigateway/lib/model.ts:
```diff
@@ -40,7 +89,107 @@ export class EmptyModel implements IModel {
+   * Generates the formatted reference to this model, so that another model can reference it, in its schema
+   */
+  public get referenceForSchema(): string {
+    return `https://apigateway.amazonaws.com/restapis/${this.restApi.restApiId}/models/${this.modelId}`;
```
Might need to use `cdk.Aws.urlSuffix` instead of `amazonaws.com` here.

john-shaskin replied:
Sure. I can change that.

RomainMuller commented on packages/@aws-cdk/aws-apigateway/lib/restapi.ts:
```diff
@@ -279,6 +280,17 @@ export class RestApi extends cdk.Construct implements IRestApi {
+   * @param schema The schema to use to transform data to one or more output formats.
+   * @param description A description that identifies this model.
+   */
+  public addModel(name: string, contentType: string, schema: any, description?: string): Model {
```
You should wrap `contentType`, `schema`, and `description?` in a "props" type object.

Also, I would advise against using `any` for the `schema`. You will not be able to export the type from `jsonschema` (if you decide to go with it) because it is not a JSII module... But I guess you could duplicate it (it's not ideal, but it's pragmatic).

john-shaskin replied:
Yeah. I was definitely trying to add that type safety to the schema before, but was running into those JSII errors. I'll try your suggestion around adding jsonschema to the bundledDependencies.

john-shaskin replied:
I was toying with creating a duplicate of the JSONschema type. The JSII linting rules were not fond of the `$schema` and `$ref` properties, as they were detected as not conforming to camelCasing. But I guess this could maybe be handled by transforming the input schema into a CFN schema object.

eladb commented on packages/@aws-cdk/aws-apigateway/README.md:
```diff
@@ -146,6 +146,67 @@ const proxy = resource.addProxy({
+const method = api.beer.addMethod('GET', getBeerLambdaHandler, {
+  methodResponses: [{
+      statusCode: '200',
+      responseModels: {
```
the model seems to know what content type it supports. Is there a more concise api that can take advantage of this (something like “addResponseModel”?

john-shaskin replied:
Yeah, that seems like a useful way to help steer people away from accidental content type mismatches.

eladb commented on packages/@aws-cdk/aws-apigateway/README.md:
````diff
@@ -146,6 +146,67 @@ const proxy = resource.addProxy({
+
+```ts
+// Direct addition to REST API
+const beerModel = api.addModel('Beer', 'application/json', {
````
Use keyword arguments instead of positional (basically extract the common options from ModelProps and reuse here)

john-shaskin replied:
Done.

eladb commented on packages/@aws-cdk/aws-apigateway/lib/model.ts:
```diff
@@ -1,5 +1,54 @@
+   * The ID of a REST API with which to associate this model.
+   * Required.
+   */
+  restApi: IRestApi,
```
Use semi-colons instead of commas

john-shaskin replied:
Done.

eladb commented on packages/@aws-cdk/aws-apigateway/lib/model.ts:
```diff
@@ -1,5 +1,54 @@
+   * Required.
+   * @see http://json-schema.org/
+   */
+  schema: any, // TODO: Validate that the schema provided matches the JSON schema spec.
```
Yes please!

john-shaskin replied:
Working on it, but JSII linting does not like the `$ref` and `$schema` properties in my JsonSchema interface. Attempting to remove those from the interface, but transform the output schema object to use $schema and $ref. It's getting kind of ugly with all the recursive properties...

john-shaskin replied:
Alright, I've got a little mapper class to take care of transforming the `ref` and `schema` properties to include the $ prefix.

eladb commented on packages/@aws-cdk/aws-apigateway/lib/model.ts:
```diff
@@ -40,7 +89,107 @@ export class EmptyModel implements IModel {
+/**
+ * Defines a reference to a model created outside of the current CloudFormation stack.
+ */
+export class ImportedModel extends cdk.Construct implements IModelConstruct {
```
Shouldn’t be exported

john-shaskin replied:
Removed the export.

eladb commented on packages/@aws-cdk/aws-apigateway/package.json:
```diff
@@ -82,7 +82,9 @@
     "exclude": [
-      "resource-attribute:@aws-cdk/aws-apigateway.IRestApi.restApiRootResourceId"
+      "resource-attribute:@aws-cdk/aws-apigateway.IRestApi.restApiRootResourceId",
+      "resource-interface-extends-construct:@aws-cdk/aws-apigateway.IModel",
```
I don’t see a good reason to exclude these

john-shaskin replied:
This ties into my explanation to @RomainMuller's comment above about the two interfaces for model. The original IModel was added to start getting model support in for MethodResponse, but only going as far as helping to allow the static Empty and Error models to be referenced. I didn't think it necessarily made sense to complicate those model types by making them implement Construct. That said, as above, I can look into what's needed to converge these into one interface. In which case, I will remove this exclusion.

john-shaskin replied:
I was originally getting build errors on the `IModel` interface, with those complaints about extending construct and including an export() method, but I don't seem to be now. So, I've removed the exclusions. 🤷‍♂️

john-shaskin commented:
@RomainMuller and @eladb, I've made an attempt to address all your requests/comments above. Most of it was pretty straight forward. However, there are a couple of things that I'm still a bit unsure of:

1. My most recent commit is an attempt to address the suggestion that `MethodReponse` objects should have an `addResponseModel` method. To do this, I've renamed the existing interface, and converted MethodResponse to a class, so that this method can be added. Exposing the content type on a model ended up feeling a bit awkward. It's easy on the concrete Model implementation, as the content type is explicitly set, when defining the model. However, if you are importing a model or trying to use the default Empty or Error models, then I wasn't sure quite what to do. In the latter case, I can perhaps assume that those default Empty and Error models are always JSON. But for imported models, it felt weird to add content type to the import prop interface.
2. On the suggestion to converge the two interfaces for Model, I've left the interfaces separate. The `IModelRef` base interface has remained separate from `IModel` largely to support referring to the Empty and Error default models, without needing to have those classes implement Construct and an export method. It didn't feel quite right to me to go that route. However, if you guys disagree, I'm happy to make the necessary changes to converge on a single `IModel` interface.
